src/Irc.hs

```diff
@@ -39,25 +39,32 @@ import           Network.IRC.CTCP             (CTCPByteString, asCTCP)
 import           Prelude                      hiding (and)
 import           System.Console.Concurrent    (outputConcurrent)
 import           System.IO.Error              (ioeGetErrorString)
+import           Control.Exception.Lifted
 
 import           Network.SimpleIRC
 
-connectTo :: IrcParams -> Bool -> IO () -> IO () -> IrcIO Connection
-connectTo params withDebug onConnected onJoined = do
+connectTo :: Options -> IrcParams -> Bool -> IO () -> IO () -> IrcIO Connection
+connectTo opt@Options{..} params withDebug onConnected onJoined = do
     (conf, bcs) <- lift $ config params
     con <- connect' conf withDebug
     lift $ mapM_ ((\f -> f con) . onConnect) (hooks params)
     lift onConnected
-    joined <- lift $ waitForAll bcs
+    joined <- lift $ catch (waitForAll bcs) (handler con)
```

Comment:
Wouldn't an exception here be handled and rethrown by the `bracket` that ultimately calls this function which then releases the connection?